### PR TITLE
DataProtection use built in AesGcm where available

### LINKED ISSUE #20689: GCM encryption not available on linux when using DataProtection
It seems that DataProtection does not support gcm on linux, I get System.PlatformNotSupportedException stating that gcm is windows only.
That happens both in aspnetcore 3.1, and according to my code reading, also in current master.
I discovered that the AuthenticatedEncryptorFactory checks if the os is windows in the gcm case and throws, but https://github.com/dotnet/runtime/blob/master/docs/design/features/cross-platform-cryptography.md seems to state that dotnet core does support gcm on non windows. I haven't tested so it is still possible it is not supported in dotnet core 3.0/3.1, though.

### PR DESCRIPTION
**Pull request by HaoK**
Fixes https://github.com/dotnet/aspnetcore/issues/20689

Use new framework crypto apis now that they are available, this enable these apis on non windows for non net461/netcore2.

### CODE REVIEW

**Comment by HaoK** [comment]
@GrabYourPitchforks for the crypto @dotnet/aspnet-build for adding a new algorithms dependency, @dotnet/aspnet-blazor-eng for fyi/review

**Review by wtgodbe — APPROVED** [review]

File: src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs
Comment (dougbu) [inline src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:79]: Why is this block still needed❔ Might need comments before these preprocessor instructions explaining the restriction.

System.Security.Cryptography.Algorithms supports TFMs going back to `net46` and `netstandard1.3`. So, is this actually about `ReadOnlySpan<byte>`❔

Comment (HaoK) [reply src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:79]: Perhaps, I'm not familiar with these apis at all, but I'll defer to @GrabYourPitchforks on the specifics, there's only Span versions of this api I believe

Comment (HaoK) [reply src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:79]: Note: intelligence also claims that the AesGcm class is only available in net6.0 and not in net461/netstandardwhich is why these blocks were everywhere

Comment (pranavkm) [reply src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:79]: Consider inverting the check. 

```
#if NETCOREAPP
...
#else
...
```

File: src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs
Comment (pranavkm) [inline src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:187]: Any reason not to do the Aes inside the `unsafe` block? You could use `vars` if you did that

File: eng/Versions.props
Comment (pranavkm) [inline eng/Versions.props:121]: The package doesn't appear to be used. Do you know why this is required?

Comment (HaoK) [reply eng/Versions.props:121]: Nope I was just doing this out of habit, I'll remove then

File: src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs
Comment (HaoK) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:185]: Nice, done

File: src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs
Comment (HaoK) [reply src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:79]: Done

File: eng/Versions.props
Comment (HaoK) [reply eng/Versions.props:121]: I think this is needed by Dependenices.props since I got a code check failure when i removed this

File: src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs
Comment (javiercn) [inline src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Does this dispose at the end of the block or at the end of the method (99% sure the answer is block, but want to make sure. It might be better to use the old using syntax to be very explicit)

File: src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs
Comment (javiercn) [inline src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:87]: ❤️ 

There are a bunch of places in data protection where we could leverage new primitives introduced to the framework in the past few releases. Might be interesting to have an issue for that. Identity likely can also benefit from it.

File: src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs
Comment (HaoK) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Yeah from the docs I believe its the block (once it goes out of scope) https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement 

I can change it to have braces if you feel strongly about it, or we can just start training ourselves to read these new usings now :)

Comment (javiercn) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: I'm fine training myself, just bring it up because it's "security" sensitive code, in any other case I wouldn't have blinked.

File: eng/Versions.props
Comment (pranavkm) [reply eng/Versions.props:121]: Do you know what the error was? Weird that a version is required here without it being used.

File: src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs
Comment (pranavkm) [inline src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:79]: Should we also update the overload that takes a byte array to use this new API?

File: eng/Versions.props
Comment (HaoK) [reply eng/Versions.props:121]: https://dev.azure.com/dnceng/public/_build/results?buildId=974536&view=logs&j=efda9e5f-cf98-536e-4181-9b6c13ac35b3&t=de5280bb-c511-517b-1959-798dcf9a84da&l=740

##[error]Missing version variable 'SystemSecurityCryptographyAlgorithmsVersion' in the 'Automated' property group in F:\workspace\_work\1\s/eng/Versions.props

Comment (HaoK) [reply eng/Versions.props:121]: Checking that Versions.props and Version.Details.xml match

So I'm guessing I would need to remove the one in version.details.xml too?

File: src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs
Comment (HaoK) [reply src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:79]: Ah yeah, I missed that good catch!

File: eng/Versions.props
Comment (pranavkm) [reply eng/Versions.props:121]: yes please. Let's remove it from both locations.

File: eng/Dependencies.props
Comment (pranavkm) [inline eng/Dependencies.props:78]:
```suggestion
```

File: src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs
Comment (pranavkm) [inline src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:108]:
```suggestion
                    bufA.AsSpan(start: offsetA, length: countA),
```

File: eng/Dependencies.props
Comment (HaoK) [reply eng/Dependencies.props:78]: Is this picked up automatically already?  AesGcm is in that package

File: src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: This code doesn't belong here. The `Cng` namespace should consist solely of p/invoke calls into BCRYPT. This probably belongs under the `Managed` namespace instead.

Comment (HaoK) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Ah, so basically all the places that are doing GCM need to have this switch instead of putting it in the Cng/Gcm, and that we should just use the ManagedEncrypter instead?

Comment (GrabYourPitchforks) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Kinda. There are four implementations of authenticated encryption in the DataProtection stack today. (See [the _ConfigurationModel_ namespace](https://github.com/dotnet/aspnetcore/tree/main/src/DataProtection/DataProtection/src/AuthenticatedEncryption/ConfigurationModel).)

1. `AuthenticatedEncryptorDescriptor`, which is a generalized "I tell you the algorithm, you choose the best platform implementation" descriptor.
2. `CngCbcAuthenticatedEncryptorDescriptor`, which is encrypt-then-MAC, specifically using Windows CNG.
3. `CngGcmAuthenticatedEncryptorDescriptor`, which is GCM, specifically using Windows CNG.
4. `ManagedAuthenticatedEncryptorDescriptor`, which is encrypt-then-MAC, specifically using .NET `SymmetricAlgorithm` and `KeyedHashAlgorithm` abstractions.

(2) and (3) are specifically tied to CNG. They allow callers to fulfill compliance requirements by selecting specific CNG implementations, including hardware-based implementations. (Take a look at all of the `[ApplyPolicy]` attributes [here](https://github.com/dotnet/aspnetcore/blob/dfe625fbd122557cc4dda0f3d6f3dc5c7afc8c8e/src/DataProtection/DataProtection/src/AuthenticatedEncryption/ConfigurationModel/CngGcmAuthenticatedEncryptorConfiguration.cs).) These code paths should never do anything other than p/invoke into bcrypt.dll.

(4) is generalized in that it uses existing .NET abstractions, but it's encrypt-then-MAC, which isn't relevant to GCM. So let's ignore it for now.

(1) is the most useful for this scenario since it's already generalized. All you need to do is specify the desired algorithm family, and it'll choose the best implementation available on the current operating system. [It already handles GCM.](https://github.com/dotnet/aspnetcore/blob/460d4d95da28d1be8ecc6418cf58a124723e8043/src/DataProtection/DataProtection/src/AuthenticatedEncryption/ConfigurationModel/AuthenticatedEncryptorDescriptor.cs#L41)

In addition to the _ConfigurationModel_ namespace mentioned previously, each of the above descriptors has an associated factory in [the _AuthenticatedEncryption_ namespace](https://github.com/dotnet/aspnetcore/tree/main/src/DataProtection/DataProtection/src/AuthenticatedEncryption). This is where you need to make your targeted change. Remove the OS check from the block below, then create a new internal type which implements _IOptimizedAuthenticatedEncryptor_ and which delegates to a backing `AesGcm` instance.

https://github.com/dotnet/aspnetcore/blob/460d4d95da28d1be8ecc6418cf58a124723e8043/src/DataProtection/DataProtection/src/AuthenticatedEncryption/AuthenticatedEncryptorFactory.cs#L55-L72

In theory that's all you need to do. Maybe with an `#ifdef` around the OS version check if you're targeting a platform where the `AesGcm` type doesn't exist. Then somebody can use [the existing builder function](https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/DataProtection/DataProtection/src/DataProtectionBuilderExtensions.cs#L602) and it should work regardless of what platform you're running on.

```cs
builder.UseCryptographicAlgorithms(new AuthenticatedEncryptorConfiguration {
    EncryptionAlgorithm = EncryptionAlgorithm.AES_256_GCM
});
```

Comment (HaoK) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Thanks @GrabYourPitchforks that information helps a lot, so I started looking working on a new `ManagedGcmAuthenticatedEncryptor` class, the main question I had was whether I needed to still do any of the logic in the existing `GcmAuthenciatedEncryptor` specifically the AesGcm apis require a nonce, and that's obviously not something in the outer Encrypt/Decrypt apis.  So I can generate a random nonce on Encrypt, but it seems like I need to then mimic the assumptions in GcmAuthenticatedEncryptor's DecryptImpl, namely the cipher payload is: 

`keyModifer || nonce || encryptedData || authenticationTag`

Just checking that my understanding is correct?

Comment (GrabYourPitchforks) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Yes, your understanding is correct. The implementations should be near-identical, except that one uses _AesGcm_ and the other uses CNG. In fact, the two implementations should have 100% compatible payloads. If you create a ManagedGcmAuthenticatedEncryptor initialized over (some_secret, AES_256_GCM), and if you then feed (some_purpose, some_plaintext) to it, the resulting ciphertext should be readable by a CngGcmAuthenticationEncryptor initialized over (same_secret, "AES", "Microsoft default CNG provider") \[or whatever the default provider name is\] and passed (same_purpose, ciphertext) as input; and vice versa.

The ManagedCbcMacAuthenticatedEncryptor and CngCbcMacAuthenticatedEncryptor behave similarly: they can decipher each other's payloads because they create the context header in the same way and they generate the ciphertexts in the same way.

[We even document how the context header is generated](https://docs.microsoft.com/aspnet/core/security/data-protection/implementation/context-headers) in an effort to help formalize this guarantee.

Comment (GrabYourPitchforks) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: As a follow-up, I'd recommend renaming the internal _GcmAuthenticatedEncryptor_ type to _CngGcmAuthenticatedEncryptor_ (to highlight that it's CNG-specific), and I'd name the new internal encryptor type _AesGcmAuthenticatedEncryptor_ (to highlight that it's `AesCng`-specific).

Comment (HaoK) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Okay I think I precomputed the 3 context headers now :)  I'm looking at the decrypt function and I'm sort of stuck on what the managed replacement for the derive key with context header call would be as I see this calls down into BCrypt's DeriveKey:

https://github.com/dotnet/aspnetcore/blob/b1a59e8c70e53dbe1784ddac6a9fea55565fa56c/src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs#L145

Comment (GrabYourPitchforks) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Check out the type *ManagedSP800_108_CTR_HMACSHA512*, which is a managed implementation of the SP800-108 algorithm. You should be able to call that routine from the new encryptor type. The _prfFactory_ parameter should be the lambda function `(kdk) => new HMACSHA512(kdk)`.

Comment (GrabYourPitchforks) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:163]: Aside: this code is all from the pre-span days. It'd be a welcome refactoring in a separate PR to start spanifying this to avoid all the unsafe goop.

Comment (HaoK) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs:164]: Okay, I added the new class and did my best to mimic the existing logic but using the style of the managed encryptor.  The round trip tests are failing that are using the CngAes -> AesGcm round trip tests, and as far as I can tell the surrounding code looks fine, I verified that the keymodifier/nonce/tags are all being split up the same and passed into AesGcm, but the decrypt is failing with an authentication tag does not match.  The subkey derivation looks fine as well, in the debugger in that the same sub key is derived in Cng and Aes.  Maybe it will be obvious to you how I'm clearly holding things wrong slightly?

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (HaoK) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:128]: This is the decrypt that fails with:  System.Security.Cryptography.CryptographicException : The computed authentication tag did not match the input authentication tag.

I verified that derviedKey, nonce, and tag match the encrypt versions in the CngGcm side

File: src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs
Comment (HaoK) [reply src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs]: Okay I found it finally, I had a bogus "-" in one of the precomputed context headers which caused the roundtrip between cng and aes to fail :(  The tests are green now including the new Aes round trip variant, so I think this is ready for review now

**Comment by HaoK** [comment]
@GrabYourPitchforks when you get a chance could you review the new managed implementation and the test I added to make sure it looks legit?

File: src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:105]: Extreme nits: No need to use unsafe here. If desired, you can also put the _NoInlining | NoOptimization_ attribute within an `#if !NETCOREAPP` block. Maybe those can come with the spanification PR so as not to cause noise here.

File: src/DataProtection/DataProtection/src/AuthenticatedEncryption/AuthenticatedEncryptorFactory.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/DataProtection/src/AuthenticatedEncryption/AuthenticatedEncryptorFactory.cs:59]: Nit: I'd recommend not duplicating the secret here, and instead doing it in the class ctor. That way, regardless of how the encryptor object is created, it always owns its own secret.

File: src/DataProtection/DataProtection/test/AuthenticatedEncryption/ConfigurationModel/AuthenticatedEncryptorDescriptorTests.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/DataProtection/test/AuthenticatedEncryption/ConfigurationModel/AuthenticatedEncryptorDescriptorTests.cs:91]: The _control_ should be an instance of `CngGcmAuthenticatedEncryptor` (see line 65 above for an example). That way the control uses CNG, and the test uses the managed `AesGcm` type. As it's written currently, both the control and the test use the managed `AesGcm` type.

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:20]: Should probably say "see the class-level comment on CngGcmAuthenticatedEncryptor".

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:27]: For these, may help to add a comment referencing the `AesGcmAuthenticatedEncryptor.CreateContextHeader` method. That way future code owners know where the magic headers came from.

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:61]: Prefer `throw CryptoUtil.Fail("some message");`. It triggers a proper assert when under a debugger.

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:122]: Nit: unsafe not needed.

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (GrabYourPitchforks) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:157]: Nit: No need to pin this buffer. It doesn't contain sensitive data.

**Review by GrabYourPitchforks — APPROVED** [review]
Only thing that really needs fixed is the comment on _AuthenticatedEncryptorDescriptorTests.cs_. Approved once that's done. Everything else is at your discretion.

Thanks for getting to this! 👍

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (pranavkm) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:14]: Super nit, you could put thiis at the top of the file (right before the usings)

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (pranavkm) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:15]:
```suggestion
    // An encryptor that uses AesGcm to do encryption
```

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (pranavkm) [inline src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:180]: @GrabYourPitchforks would it be unwise to use the shared ArrayPool here?

File: src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs
Comment (HaoK) [reply src/DataProtection/Cryptography.Internal/src/CryptoUtil.cs:102]: Will address this in the span PR

File: src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs
Comment (HaoK) [reply src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:180]: Going to merge this as is, can make this switch in the follow up / spanification PR if needed

Comment (GrabYourPitchforks) [reply src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs:180]: Correct, because it's a cryptographic secret. Within the runtime, our crypto libraries use an internal [`CryptoPool`](https://github.com/dotnet/runtime/blob/release/5.0/src/libraries/Common/src/System/Security/Cryptography/CryptoPool.cs) type instead of the public singleton `ArrayPool<T>.Shared` instance.

During the spanification work, this can be switched to using _stackalloc_ (the common case) or the pinned object heap. Then no pooling will be needed at all.

**Comment by GrabYourPitchforks** [comment]
I finished reviewing again through the commit https://github.com/dotnet/aspnetcore/pull/29814/commits/a752598586163c9f7d2938b7e43c207e0f6b3ee7 and everything looks great. Thanks so much Hao!

**Pull request merged**
